File: JayKay24/Stacks and Queues 1/min-stack.ts
class MinStack {
  private stack: number[];
  private minValTracker: number[];

  constructor() {
    this.stack = [];
    this.minValTracker = [];
  }
  /**
   * Pushes the element val onto the stack
   *
   * @param val - the value to push onto the stack
   */
  push(val: number): void {
    if (
      this.minValTracker.length === 0 ||
      val <= this.minValTracker[this.minValTracker.length - 1]
    ) {
      this.minValTracker.push(val);
    }

    this.stack.push(val);
  }

  /**
   * Removes the element on top of the stack
   */
  pop(): void {
    if (this.stack.length > 0) {
      if (
        this.minValTracker[this.minValTracker.length - 1] ===
        this.stack[this.stack.length - 1]
      ) {
        this.minValTracker.pop();
      }
      this.stack.pop();
    }
  }

  /**
   * Gets the top element of the stack
   *
   * @returns The top element of the stack
   */
  top(): number {
    return this.stack[this.stack.length - 1];
  }

  /**
   * Retrieves the minimum element in the stack
   *
   * @returns The minimum element in the stack
   */
  getMin(): number {
    return this.minValTracker[this.minValTracker.length - 1];
  }
}

File: stacks-and-queues-1/MinStack.java
import java.util.*;

public class MinStack {
    Stack<Integer> numbers;
    Integer minValueTracker;

    MinStack() {
        numbers = new Stack<>();
    }

    void getMin(){
        if (numbers.isEmpty()){
            System.out.println("Stack is empty");
        } else {
            System.out.println("Min element is: " + minValueTracker);
        }
    }

    void push(Integer integer){
        if (numbers.isEmpty()){
            minValueTracker = integer;
            numbers.push(integer);
            return;
        }

        if (integer < minValueTracker){
            numbers.push((integer * 2) - minValueTracker);
            minValueTracker = integer;
        } else {
            numbers.push(integer);
            System.out.println("Number inserted");
        }
    }

    void pop(){
        if (numbers.isEmpty()){
            System.out.println("Stack is empty");
            return;
        }

        System.out.println("Removed top element");
        Integer top = numbers.pop();

        if (top < minValueTracker){
            System.out.println(minValueTracker);
            minValueTracker = (minValueTracker * 2) - top;
        } else {
            System.out.println(top);
        }
    }

    Integer top(){
        if (!(this.numbers.size() == 0)){
            return this.minValueTracker;
        }
        return minValueTracker;
    }
}